dashboard/test/lib/services/user/password_resetter_by_email_test.rb

```diff
@@ -0,0 +1,154 @@
+
+        _(mail).wont_be_nil
+        _(mail.to).must_equal [email]
+        _(mail.subject).must_equal 'Code.org reset password instructions'
+      end
+    end
+
+    context 'when account is an lti account' do
+      let!(:user) {create(:teacher, :with_lti_auth)}
+      # Email is changed by lti auth option, so we need to reference that email
+      let!(:email) {user.email}
+
+      context 'when account does not have email authentication' do
+        it 'tracks PasswordResetEmailAuthNotFound metric' do
+          expected_env = 'expected_env'
+
+          CDO.expects(:rack_env).returns(expected_env)
+          Cdo::Metrics.
+            expects(:put).
+            with('User', 'PasswordResetEmailAuthNotFound', 1, {Environment: expected_env}).
+            once
+
+          reset_password
+        end
+
+        it 'returns new user instance without reset password tokens and does not send email' do
+          _reset_password.must_be_instance_of User
+          _(reset_password.reset_password_token).must_be_nil
+          _(reset_password.reset_password_sent_at).must_be_nil
+          _(reset_password.raw_token).must_be_nil
+
+          _(mail).must_be_nil
+        end
+      end
+
+      context 'when account has email authentication' do
+        let(:email_auth_option) {create(:authentication_option, email: email)}
+
+        before do
+          user.authentication_options.append(email_auth_option)
+        end
+
+        it 'returns user with generated reset password tokens and sends an email' do
+          _reset_password.must_equal user
+          _(reset_password.reset_password_token).wont_be_nil
+          _(reset_password.reset_password_sent_at).wont_be_nil
+          _(reset_password.raw_token).wont_be_nil
+
+          _(mail).wont_be_nil
+          _(mail.to).must_equal [email]
+          _(mail.subject).must_equal 'Code.org reset password instructions'
+        end
+      end
+    end
+
+    context 'when account is a google account' do
+      let!(:user) {create(:teacher, :with_google_authentication_option, email: email)}
+
+      context 'when account does not have email authentication' do
+        before do
+          user.authentication_options.find_by(credential_type: 'email').destroy
+        end
+
+        it 'tracks PasswordResetEmailAuthNotFound metric' do
+          expected_env = 'expected_env'
+
+          CDO.expects(:rack_env).returns(expected_env)
+          Cdo::Metrics.
+            expects(:put).
+            with('User', 'PasswordResetEmailAuthNotFound', 1, {Environment: expected_env}).
+            once
+
+          reset_password
+        end
+
+        it 'returns new user instance without reset password tokens and does not send email' do
+          _reset_password.must_be_instance_of User
+          _(reset_password.reset_password_token).must_be_nil
+          _(reset_password.reset_password_sent_at).must_be_nil
+          _(reset_password.raw_token).must_be_nil
+
+          _(mail).must_be_nil
```

Comment:
Just a note: if you feel the need to add 'and ...' to the test case, it's probably a sign that you should create a separate test case for it.
https://www.betterspecs.org/#single